### ISSUE TITLE
Using RTD's "Version warning" feature

### ISSUE THREAD

**Issue by amotl** [crate/crate-docs-theme#323]
@amotl maybe a bit late to the party, but can't #322 be resolved via https://docs.readthedocs.io/en/stable/versions.html#version-warning ?

_Originally posted by @msbt in https://github.com/crate/crate-docs-theme/issues/322#issuecomment-1064085865_

**Comment by amotl** [comment]
Dear @msbt,

thanks for your suggestion. I just checked the projects [1,2] and found the corresponding feature to be enabled already.

![image](https://user-images.githubusercontent.com/453543/157697007-e463cec5-6062-4ea7-b63e-b8183ee0556f.png)

Now, we might want to investigate why it's not working. On this matter, I discovered this output on the Browser console. Maybe it is the culprit?

![image](https://user-images.githubusercontent.com/453543/157696839-304e8573-fcdf-40fa-b1e9-880067833080.png)

With kind regards,
Andreas.

[1] https://readthedocs.org/dashboard/crate/advanced/
[2] https://readthedocs.org/dashboard/crash/advanced/

**Comment by msbt** [comment]
@amotl I don't know to be honest, those errors are appearing since we got rid of our custom docutils (!) and used the default one.

**Comment by msbt** [comment]
I think it was around that time [here](https://github.com/crate/crate-docs-theme/commit/a0777ba59e31adf296868934c47b56e554d74659) or rather [this one](https://github.com/crate/crate-docs-theme/commit/264619b062930a50d84f43d3e5db0b7d92d67f05) when you/we added all the webpack stuff. Before that we had a custom doctools.js that shipped with the theme I believe

**Comment by msbt** [comment]
Hah, beat me to it, yes, at least I think that was the time when the error showed up

**Comment by amotl** [comment]
I can see `doctools.js` being referenced, but I can't find it back. Neither the file, nor the commit which removes it.

**Comment by msbt** [comment]
Hrm odd, maybe I'm remembering it wrong, but I'm pretty sure it was around that time. I'll check back on Monday if I can find something more concrete, sorry for the wild goosechase :/

**Comment by msbt** [comment]
@amotl ok maybe that wasn't it, but we do have one repo with an old template online, I'll compare a current one with https://crate.io/docs/crate/reference/en/3.3/

**Comment by jayeff** [comment]
@msbt @amotl I opened a PR which would upgrade docs-theme for crate 3.3 branch: https://github.com/crate/crate/pull/12209 Preview is here: https://crate--12209.org.readthedocs.build/en/12209/

Just upgrading docs-theme alone does not seem to fix the version-warning though

**Comment by amotl** [comment]
Hi again,

I just looked up the corresponding implementation by RTD and wanted to share corresponding pointers. They can be useful to find out why it doesn't work on our pages.

- Version comparison and evaluation of the `is_highest` flag happens within [views/footer_views.py](https://github.com/readthedocs/readthedocs.org/blob/7.4.1/readthedocs/api/v2/views/footer_views.py#L30-L83).
- The banner is rendered by [js/doc-embed/version-compare.js](https://github.com/readthedocs/readthedocs.org/blob/7.4.1/readthedocs/core/static-src/core/js/doc-embed/version-compare.js).
- In the footer, other details about versioned documenation is handled, but nothing about the version warning / `is_highest` flag. See [js/doc-embed/footer.js](https://github.com/readthedocs/readthedocs.org/blob/7.4.1/readthedocs/core/static-src/core/js/doc-embed/footer.js).

I hope this will help to give us a trace to follow.

With kind regards,
Andreas.

**Comment by amotl** [comment]
Hi again,

@msbt found that it works perfectly on the `readthedocs.io` domain. The RTD-native version warning admonition box can be inspected on https://crate.readthedocs.io/en/4.6/ and looks like 

![image](https://user-images.githubusercontent.com/453543/158373571-3588aaba-ac80-403e-9c39-77c51e225257.png)

@seut: If you don't have any objections, I would prefer using this solution instead of our own thing #322. @msbt continues to trace the problem outlined at https://github.com/crate/crate-docs-theme/issues/323#issuecomment-1064193967, which currently happens to render all RTD injections defunct on the crate.io domain but told me that he might already have a clue.

With kind regards,
Andreas.

**Comment by seut** [comment]
Great! Yeah definitely prefer this as well. Thanks a lot for investigating @amotl @msbt

**Comment by amotl** [comment]
Hi again,

@msbt shared some insights about his investigations with me, thanks a stack.

- At [1], we can see the variable `proxied_api_host` is responsible for designating the right URI to RTD's API endpoint. By default, it is configured as `/_`, see [2]. This will have to be adjusted appropriately on a per-project level.
- At the same time, we will have to take care that the Nginx reverse proxy configuration will pass through those requests correctly. Currently, this is happening at [crateio_docs/rewrites/02_versions.conf#L48-L49](https://github.com/crate/infrastructure/blob/5d758d76af1305ed98979e532faee7581c1d52d7/playbooks/roles/web/files/crateio_docs/rewrites/02_versions.conf#L48-L49).

With kind regards,
Andreas.

[1] https://github.com/readthedocs/readthedocs.org/blob/7.4.1/readthedocs/core/static-src/core/js/doc-embed/footer.js#L67
[2] https://test-builds-webhook.readthedocs.io/en/latest/conf.html

**Comment by amotl** [comment]
A full `_/api` call, for getting the footer RTD HTML inject, looks like

https://crate.readthedocs.io/_/api/v2/footer_html/?callback=callback&project=crate&version=latest&page=index&theme=crate&format=jsonp&docroot=/docs/&source_suffix=.rst

All `_/api` URLs must at least satisify, for example, this URL scheme, in order to deliver the corresponding RTD HTML inject code.

https://crate.readthedocs.io/_/api/v2/footer_html/?project=crate&version=latest

All other variants, like omitting the trailing slash, or omitting one of the required parameters, will make the API respond with `404 Not Found`. So, the rewriting applied by Nginx must exactly match/reproduce this URL pattern.

**Comment by amotl** [comment]
Hi again,

I hope that two patches will help to improve the situation.

### Improve Nginx configuration
The first patch, https://github.com/crate/infrastructure/pull/2850, relates to appropriate Nginx configuration to forward query string parameters to an upstream HTTP server using `proxy_pass` [^1].

```diff
-    proxy_pass https://example.readthedocs.io:443/$1;
+    proxy_pass https://example.readthedocs.io:443/$1$is_args$args;
```

Reference/demonstration projects/URLs:

- [RTD inject asset for `crate-tutorials`](https://crate.io/docs/crate/tutorials/_/api/v2/footer_html/?project=crate-tutorials&version=latest) (does not work yet)
- [RTD inject asset for `crate-howtos`](https://crate.io/docs/crate/howtos/_/api/v2/footer_html/?project=crate-howtos&version=latest) (works, after adjusting the Nginx configuration)

### Improve RTD project configuration
The second patch, https://github.com/crate/crate-docs-theme/pull/326, attempts to adjust RTD's `proxied_api_host` context variable on a per-project level, to make the project use an URL like outlined above for `crate-howtos`. This code has not been verified yet.

With kind regards,
Andreas.

[^1]: https://stackoverflow.com/a/8130872

**Comment by amotl** [comment]
Hi again,

the improvements above have been verified on behalf of the `crash` project. There are three indicators where we can see it works now.

### Indicator I
You can see that it works by visiting the most recent documentation version at https://crate.io/docs/crate/crash/en/0.27/, where the RTD-native version chooser is now active at the bottom right corner. We probably will want to hide this again at the JavaScript or CSS level. I had already discussed this detail with @msbt.

![image](https://user-images.githubusercontent.com/453543/160172777-bea67915-8cb0-4ece-b24a-36c2a7590a2e.png)

### Indicator II

On the other two documentation versions, after rebuilding them, the **version warning admonition** box is now displayed correctly.

- https://crate.io/docs/crate/crash/en/0.25/
- https://crate.io/docs/crate/crash/en/0.26/

![image](https://user-images.githubusercontent.com/453543/160173800-1b570d1b-59d7-40c1-90de-dee560d48d86.png)


### Indicator III

At https://readthedocs.org/dashboard/crash/traffic-analytics/, we can see that the RTD-native analytics tracking also started working.

![image](https://user-images.githubusercontent.com/453543/160176747-8348b656-a68a-4dcc-b77d-65a70d624207.png) ![image](https://user-images.githubusercontent.com/453543/160176814-324ea678-ab8e-4744-b336-58ea433033f1.png)


With kind regards,
Andreas.

**Comment by msbt** [comment]
Nice one @amotl, looking good! Now we need to figure out how to block/hide that injected readthedocs box at the bottom right, because all the links it contains are on rtd and not crate.iio (i.e. https://crash.readthedocs.io/en/latest/ and https://crash.readthedocs.io/en/0.27/). 

Mika started a discussion about that [a while back](https://github.com/readthedocs/readthedocs.org/issues/3233) but I don't think this is possible out of the box just yet. It's a shame though, because with that box also come quite a few assets (including fontawesome) which waste traffic if we don't use them.

**Comment by amotl** [comment]
> looking good

I like it as well.

> Now we need to figure out how to block/hide that injected readthedocs box at the bottom right.

Exactly.

> @autophagy started a discussion about that [the other day].

Thank you for refering to this discussion. I think the comment by @humitos at https://github.com/readthedocs/readthedocs.org/issues/3233#issuecomment-418664393 gives some hints how to hide that element again:

>> Read the Docs injects the HTML returned by the `/footer_html` API call into `div.rst-other-versions`. If you remove that element from your theme, RTD won't be able to inject it.

> It's a shame though, because with that box also come quite a few assets (including fontawesome) which waste traffic if we don't use them.

I see. Thanks for outlining this. We might want to take care within one of the next iterations.

**Comment by msbt** [comment]
> > Read the Docs injects the HTML returned by the /footer_html API call into div.rst-other-versions. If you remove that element from your theme, RTD won't be able to inject it.

That's the thing, I don't think this is true (anymore). It creates a top-level div.injected before the `</body>` tag, not much we could remove template-wise to avoid this.

**Comment by autophagy** [comment]
The way I got around disabled this in my own themes is by adding a CSS rule of:

```
.injected {
    display: none;
}
```

since the injected RTD html is wrapped in a div with class `injected`. Unfortunately, like @msbt said, I don't think this approach prevents resources being pulled over the network for it, but that's the best I could find.

**Comment by amotl** [comment]
Dear @autophagy,

thanks for your suggestion, @msbt implemented this with 8a46277f.

Other than this, I also closed #322 because the patches outlined within this discussion enabled us to use the RTD-native version admonition box, see https://github.com/crate/crate-docs-theme/issues/323#issuecomment-1079274537. Following that, I am also closing this issue.

Thanks to everyone who participated.

With kind regards,
Andreas.
